### PR TITLE
Album purchases

### PR DESCRIPTION
**Pull request by fede1608**
## Summary
- POST /albums/:id implemented
- Tests & Docs

## Trello Card

[https://trello.com/c/nITMmUdB/31-comprar-album](https://trello.com/c/nITMmUdB/31-comprar-album)

### CODE REVIEW

File: app/controllers/albums.js [hunk 1]
```diff
@@ -11,5 +13,26 @@ module.exports = {
...
+
+  buy(req, res, next) {
+    const purchase = {
+      userId: req.user.id,
+      albumId: req.params.id
+    };
+    Purchase.findOne({ where: purchase })
+      .then(p => {
+        if (p) {
+          next(albumAlreadyPurchased);
```
Comment (mpizzagalli) [inline app/controllers/albums.js]: you can use `return next(album...` and avoid the else statement

File: app/errors.js [hunk 2]
```diff
@@ -6,10 +6,10 @@ const internalError = (message, internalCode) => ({
 exports.DEFAULT_ERROR = 'default_error';
 exports.defaultError = message => internalError(message, exports.DEFAULT_ERROR);
 
-// exports.userNotFound = {
-//   message: 'The user was not found.',
-//   internalCode: 'user_not_found'
-// };
+exports.albumAlreadyPurchased = {
```
Comment (mpizzagalli) [inline app/errors.js:9]: remember you could use the internalError function and create some more customizable messages like using params like the id.

File: app/models/purchase.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        type: DateTypes.INTEGER,
+        allowNull: false
+      },
+      date: {
+        type: DateTypes.DATE,
+        defaultValue: DateTypes.NOW
+      }
+    },
+    {
+      timestamps: false
```
Comment (mpizzagalli) [inline app/models/purchase.js]: I would leave the timestamps in true, so you can have more idea of what happened for each row.

File: migrations/migrations/20181025175848-albums.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        autoIncrement: true,
+        primaryKey: true,
+        unique: true
+      },
+      userId: {
+        type: Sequelize.INTEGER,
+        foreignKey: true,
+        allowNull: false
+      },
+      albumId: {
```
Comment (mpizzagalli) [inline migrations/migrations/20181025175848-albums.js]: usually the convention is that the db fields go in `snake_case` but js like stuff in `camelCase`, so you can leave here `album_id` and in the model leave the attribute `albumId` but with the configuration `field: 'album_id'` to bind both field and attribute

File: app/controllers/albums.js [hunk 5]
```diff
@@ -11,5 +13,26 @@ module.exports = {
...
       });
     });
+  },
+
+  buy(req, res, next) {
+    const purchase = {
+      userId: req.user.id,
+      albumId: req.params.id
+    };
+    Purchase.findOne({ where: purchase })
```
Comment (mpizzagalli) [inline app/controllers/albums.js]: We aren't checking if the album actually exists in the album provider.

File: test/albums.js [hunk 6]
```diff
@@ -54,3 +54,58 @@ describe('/albums GET', () => {
...
+      .send({
+        email: 'federico.casares@wolox.com.ar',
+        password: '12345678'
+      })
+      .then(response => {
+        chai
+          .request(server)
+          .post('/albums/1')
+          .set('authorization', `Bearer ${response.body.token}`)
+          .end((err, res) => {
```
Comment (mpizzagalli) [inline test/albums.js:62]: let's use promises instead of callbacks.

Also we are only checking the endpoint response and not if the logic is fine.

**Review by mpizzagalli — CHANGES_REQUESTED** [review]

File: app/controllers/albums.js (hunk 1 above)
Comment (fede1608) [reply app/controllers/albums.js]: fixed

File: app/models/purchase.js (hunk 3 above)
Comment (fede1608) [reply app/models/purchase.js]: changed

File: migrations/migrations/20181025175848-albums.js (hunk 4 above)
Comment (fede1608) [reply migrations/migrations/20181025175848-albums.js]: But the db is not configured to be snake or underscore, should I change it?

File: app/controllers/albums.js (hunk 5 above)
Comment (fede1608) [reply app/controllers/albums.js]: fixed

File: test/albums.js (hunk 6 above)
Comment (fede1608) [reply test/albums.js:62]: Will refactor that in another pr.

File: migrations/migrations/20181025175848-albums.js (hunk 4 above)
Comment (fede1608) [reply migrations/migrations/20181025175848-albums.js]: Fixed, but couldn't link userId to the user_id column in the foreign key column

File: test/albums.js (hunk 6 above)
Comment (fede1608) [reply test/albums.js:62]: Done in #15

**Review by mpizzagalli — APPROVED** [review]

**Pull request merged**